File: python/exercise3/order_processor.py
from decimal import Decimal, ROUND_HALF_UP
from dataclasses import dataclass

@dataclass
class OrderResult:
    is_success: bool
    final_amount: Decimal
    message: str

class OrderProcessor:
    def process_order(self, amount: Decimal, state: str, is_member: bool, loyalty_years: int) -> OrderResult:
        res = OrderResult(is_success=False, final_amount=Decimal("0.00"), message="")

        if amount > Decimal("0.00"):
            if state in ("OH", "AZ", "TX"):
                tax = Decimal("0.00")
                if state == "OH":
                    tax = amount * Decimal("0.0575")
                elif state == "AZ":
                    tax = amount * Decimal("0.056")
                else:
                    tax = amount * Decimal("0.0625")

                disc = Decimal("0.00")
                if is_member:
                    if loyalty_years >= 5:
                        disc = amount * Decimal("0.15")
                    elif loyalty_years >= 2:
                        disc = amount * Decimal("0.05")
                    else:
                        disc = amount * Decimal("0.02")

                calc = amount + tax - disc
                if calc < Decimal("0.00"):
                    calc = Decimal("0.00")

                res.is_success = True
                res.final_amount = calc.quantize(Decimal("0.01"), rounding=ROUND_HALF_UP)
                res.message = "Processed"
            else:
                res.is_success = False
                res.final_amount = Decimal("0.00")
                res.message = "Unsupported State"
        else:
            res.is_success = False
            res.final_amount = Decimal("0.00")
            res.message = "Invalid Amount"

        return res

File: exercises/javascript/exercise3/orderProcessor.js
class OrderProcessor {
  processOrder(amount, state, isMember, loyaltyYears) {
    const res = { isSuccess: false, finalAmount: 0, message: '' };

    if (amount > 0) {
      if (state === 'OH' || state === 'AZ' || state === 'TX') {
        let tax = 0;
        if (state === 'OH') {
          tax = amount * 0.0575;
        } else if (state === 'AZ') {
          tax = amount * 0.056;
        } else {
          tax = amount * 0.0625;
        }

        let disc = 0;
        if (isMember) {
          if (loyaltyYears >= 5) {
            disc = amount * 0.15;
          } else if (loyaltyYears >= 2) {
            disc = amount * 0.05;
          } else {
            disc = amount * 0.02;
          }
        }

        let calc = amount + tax - disc;
        if (calc < 0) {
          calc = 0;
        }

        res.isSuccess = true;
        res.finalAmount = Math.round(calc * 100) / 100;
        res.message = 'Processed';
      } else {
        res.isSuccess = false;
        res.finalAmount = 0;
        res.message = 'Unsupported State';
      }
    } else {
      res.isSuccess = false;
      res.finalAmount = 0;
      res.message = 'Invalid Amount';
    }

    return res;
  }
}

module.exports = OrderProcessor;
